ivadomed/losses.py

```diff
@@ -114,3 +114,103 @@ def forward(self, input, target):
         denominator = ((input + target).sum(-1) * class_weights).sum()
 
         return 1. - 2. * intersect / denominator.clamp(min=self.epsilon)
+
+
+class TverskyLoss(nn.Module):
+    """Tversky Loss.
+
+    Compute the Tversky loss defined in:
+        Sadegh et al. (2017) Tversky loss function for image segmentation using 3D fully convolutional deep networks
+
+    """
+
+    def __init__(self, alpha=0.7, beta=0.3, smooth=1.0):
+        """
+        Args:
+            alpha (float): weight of false positive voxels
+            beta  (float): weight of false negative voxels
+            smooth (float): epsilon to avoid division by zero, when both Numerator and Denominator of Tversky are zeros
+
+        Notes:
+            - setting alpha=beta=0.5: equivalent to DiceLoss
+            - default parameters were suggested by https://arxiv.org/pdf/1706.05721.pdf
+        """
+        super(TverskyLoss, self).__init__()
+        self.alpha = alpha
+        self.beta = beta
+        self.smooth = smooth
+
+    def tversky_index(self, y_pred, y_true):
+        """Compute Tversky index
+
+        Args:
+            y_pred (torch Tensor): prediction
+            y_true (torch Tensor): target
+
+        Returns:
+            float: Tversky index
+        """
+        # Compute TP
+        tp = torch.sum(y_true * y_pred)
+        # Compute FN
+        fn = torch.sum(y_true * (1 - y_pred))
+        # Compute FP
+        fp = torch.sum((1 - y_true) * y_pred)
+        # Compute Tversky for the current class, see Equation 3 of the original paper
+        numerator = tp + self.smooth
+        denominator = tp + self.alpha * fp + self.beta * fn + self.smooth
+        tversky_label = numerator / denominator
+        return tversky_label
+
+    def forward(self, input, target):
+        n_classes = input.shape[1]
+        tversky_sum = 0.
+
+        # TODO: Add class_of_interest?
+        for i_label in range(n_classes):
```

Comment:
I am wondering here: why do we have to loop? The operations in tversky_index (substraction, addition and division) can be done on multichannel images, no?

Reply:
Because in the FocalTverskyLoss --> we apply the gamma to (1-tversky_class_i) --> we need to compute them per label --> that's why I created this loop.
I could re-write this with array manipulation if you think it would be better.